cardano-cli/src/Cardano/CLI/Shelley/Run/Validate.hs

```diff
@@ -0,0 +1,359 @@
+    -> (StakeAddress, Lovelace, BuildTxWith BuildTx (Witness WitCtxStake era))
+  convert (sAddr, ll, mScriptWitnessFiles) =
+    case mScriptWitnessFiles of
+      Just sWit -> do
+        (sAddr, ll, BuildTxWith $ ScriptWitness ScriptWitnessForStakeAddr sWit)
+      Nothing -> (sAddr,ll, BuildTxWith $ KeyWitness KeyWitnessForStakeAddr)
+
+newtype TxCertificatesValidationError
+  = TxCertificatesValidationNotSupported AnyCardanoEra
+  deriving Show
+
+instance Error TxCertificatesValidationError where
+  displayError (TxCertificatesValidationNotSupported e) =
+    "Transaction certificates are not supported in " <> Text.unpack (renderEra e)
+
+validateTxCertificates
+  :: forall era.
+     CardanoEra era
+  -> [(Certificate, Maybe (ScriptWitness WitCtxStake era))]
+  -> Either TxCertificatesValidationError (TxCertificates BuildTx era)
+validateTxCertificates _ [] = return TxCertificatesNone
+validateTxCertificates era certsAndScriptWitnesses =
+  case certificatesSupportedInEra era of
+    Nothing -> Left $ TxCertificatesValidationNotSupported
+                    $ getIsCardanoEraConstraint era
+                    $ AnyCardanoEra era
+    Just supported -> do
+      let certs = map fst certsAndScriptWitnesses
+          reqWits = Map.fromList $ mapMaybe convert certsAndScriptWitnesses
+      return $ TxCertificates supported certs $ BuildTxWith reqWits
+  where
+   -- We get the stake credential witness for a certificate that requires it.
+   -- NB: Only stake address deregistration and delegation requires
+   -- witnessing (witness can be script or key)
+   deriveStakeCredentialWitness
+     :: Certificate
+     -> Maybe StakeCredential
+   deriveStakeCredentialWitness cert = do
+     case cert of
+       StakeAddressDeregistrationCertificate sCred -> Just sCred
+       StakeAddressDelegationCertificate sCred _ -> Just sCred
+       _ -> Nothing
+
+   convert
+     :: (Certificate, Maybe (ScriptWitness WitCtxStake era))
+     -> Maybe (StakeCredential, Witness WitCtxStake era)
+   convert (cert, mScriptWitnessFiles) = do
+     sCred <- deriveStakeCredentialWitness cert
+     case mScriptWitnessFiles of
+       Just sWit -> do
+         Just ( sCred
+              , ScriptWitness ScriptWitnessForStakeAddr sWit
+              )
+       Nothing -> Just (sCred, KeyWitness KeyWitnessForStakeAddr)
+
+newtype TxProtocolParametersValidationError
+  = ProtocolParametersNotSupported AnyCardanoEra
+  deriving Show
+
+instance Error TxProtocolParametersValidationError where
+  displayError (ProtocolParametersNotSupported e) =
+    "Transaction protocol parameters are not supported in " <> Text.unpack (renderEra e)
+
+validateProtocolParameters
+  :: CardanoEra era
+  -> Maybe ProtocolParameters
+  -> Either TxProtocolParametersValidationError (BuildTxWith BuildTx (Maybe ProtocolParameters))
+validateProtocolParameters _ Nothing = return (BuildTxWith Nothing)
+validateProtocolParameters era (Just pparams) =
+    case scriptDataSupportedInEra era of
```

Comment:
What's the reason we case match on `scriptDataSupportedInEra` in `validateProtocolParameters`?